docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,83 @@
+# RFC 016: Node Architecture
+
+## Changelog
+
+- April 8, 2022: Initial draft (@cmwaters)
+- April 15, 2022: Incorporation of feedback
+
+## Abstract
+
+The `node` package is the entry point into the Tendermint codebase, used both by the command line and programatically to create the nodes that make up a network. The package has suffered the most from the evolution of the codebase, becoming bloated as developers clipped on their bits of code here and there to get whatever feature they wanted working.
+
+The decisions made at the node level have the biggest impact to simplifying the protocols within them, unlocking better internal designs and making Tendermint more intuitive to use and easier to understand from the outside. Work, in minor increments, has already begun on this section of the codebase. This document exists to spark forth the necessary discourse in a few related areas that will help the team to converge on the long term makeup of the node.
+
+## Discussion
+
+The following is a list of points of discussion around the architecture of the node:
+
+### Dependency Tree
+
+The node object is currently stuffed with every component that possibly exists within Tendermint. In the constructor, all objects are built and interlaid with one another in some awkward dance. My guiding principle is that the node should only be made up of the components that it wants to have direct control of throughout its life. The node is a service which currently has the purpose of starting other services up in a particular order and stopping them all when commanded to do so. However, there are many services which are not direct dependents i.e. the mempool and evidence services should only be working when the consensus service is running. I propose to form more of a hierarchical structure of dependents which forces us to be clear about the relations that one component has to the other. More concretely, I propose the following dependency tree:
+
+![node dependency tree](./images/node-dependency-tree.svg)
```

Comment:
I like the picture, but just wonder at which level PEX should be. It is actually the first service to be started, and if I understood correctly, should also be the main switch/router that allows modules and nodes to interact.

I already had this discussion when writing a paper on Tendermint, where PEX is a transport layer or a module. While for the sake of representation and modularity, one can see it as the bottom layer of Tendermint, which interacts with the network, it is also a protocol (so, intermediate level in the hierarchy) and, due to its relevance for bootstrapping, maybe even a module/"process" by itself.

Reply:
PEX, at least how I see it, is disconnected in functionality to the router. The router dials and accepts connections to peers and manages the flow of messages between connected peers and reactors. PEX is just used to build up an address book of peers in the network. The p2p router, as a `Service` is the first to be started but PEX doesn't actually need to be started at all if the node already has a sufficiently sized address book